OpenRA.Mods.RA/Widgets/ProductionPaletteWidget.cs

```diff
@@ -203,26 +235,25 @@ public void RefreshIcons()
 			icons = new Dictionary<Rectangle, ProductionIcon>();
 			if (CurrentQueue == null)
 			{
-				if (IconCount != 0)
+				if (DisplayedIconCount != 0)
 				{
-					OnIconCountChanged(IconCount, 0);
-					IconCount = 0;
+					OnIconCountChanged(DisplayedIconCount, 0);
+					DisplayedIconCount = 0;
 				}
 
 				return;
 			}
 
-			var allBuildables = CurrentQueue.AllItems().OrderBy(a => a.Traits.Get<BuildableInfo>().BuildPaletteOrder);
-
-			var oldIconCount = IconCount;
-			IconCount = 0;
+			var oldIconCount = DisplayedIconCount;
+			DisplayedIconCount = 0;
 
 			var ks = Game.Settings.Keys;
 			var rb = RenderBounds;
-			foreach (var item in allBuildables)
+
+			foreach (var item in AllBuildables.Skip(IconRowOffset * Columns).Take(MaxRows * Columns))
```

Comment:
This causes hotkeys to only be assigned for visible actors, and changes the hotkey for a specific actor when it is scrolled.